src/Aws/Kinesis/Client/Producer.hs

```diff
@@ -354,111 +318,101 @@ generatePartitionKey gen =
   Kin.partitionKey (T.pack name)
     & either (error ∘ T.unpack) id
 
--- | This will take up to @n@ items from a 'TBMQueue'.
+-- | This will take up to @n@ items from a 'TBMQueue', with a timeout.
 --
-takeTBMQueue
-  ∷ Int
+takeTBMQueueWithTimeout
+  ∷ ∀ α
+  . Int -- ^ number of items to get
+  → Int -- ^ timeout in microseconds
   → TBMQueue α
-  → STM [α]
-takeTBMQueue n q
-  | n <= 0 = return []
-  | otherwise = do
-      res ← tryReadTBMQueue q
-      case res of
-        Just (Just x) → (x:) <$> takeTBMQueue (n - 1) q
-        _ → return []
+  → IO [α]
+takeTBMQueueWithTimeout n timeout q = do
+  timedOutVar ← newEmptyTMVarIO
+
+  let
+    timeoutAction = do
+      threadDelay timeout
+      atomically $ putTMVar timedOutVar ()
+
+    go xs
+      | length xs ≥ n = return xs
+      | otherwise = do
+          res ← Left <$> readTBMQueue q <|> Right <$> readTMVar timedOutVar
+          case res of
+            Left mx → maybe (return xs) (go ∘ (:xs)) mx
+            Right () → return xs
+
+  withAsync timeoutAction $ \h → do
+    xs ← atomically $ go []
+    xs <$ cancel h
 
 -- | A policy for chunking the contents of the message queue.
 --
 data ChunkingPolicy
   = ChunkingPolicy
-  { _cpMaxChunkSize ∷ Int
+  { _cpMaxChunkSize ∷ !Int
   -- ^ The largest chunk size that is permitted.
 
-  , _cpThrottlingDelay ∷ Int
-  -- ^ The time after which a chunk should be committed, even if the maximum
-  -- chunk size has not yet been reached.
+  , _cpMinChunkingInterval ∷ !Int
+  -- ^ The time in microseconds after which a chunk should be committed, even
+  -- if the maximum chunk size has not yet been reached.
   }
 
 -- | A 'Source' that reads chunks off a bounded STM queue according some
 -- 'ChunkingPolicy'.
 --
 chunkedSourceTBMQueue
-  ∷ ( MonadIO m
-    , MonadBaseControl IO m
-    )
-  ⇒ ChunkingPolicy
+  ∷ ChunkingPolicy
   → TBMQueue α
-  → Source m [α]
-chunkedSourceTBMQueue bp@ChunkingPolicy{..} q = do
-  terminateNow ← liftIO ∘ atomically $ isClosedTBMQueue q
-  unless terminateNow $ do
-    items ← liftIO ∘ atomically $ takeTBMQueue _cpMaxChunkSize q
+  → Source IO [α]
+chunkedSourceTBMQueue cp@ChunkingPolicy{..} q = do
+  shouldTerminate ←
+    liftIO ∘ atomically $
+      (&&) <$> isClosedTBMQueue q <*> isEmptyTBMQueue q
+
+  unless shouldTerminate $ do
+    items ← lift $ takeTBMQueueWithTimeout _cpMaxChunkSize _cpMinChunkingInterval q
```

Comment:
OK, here is the proposed implementation:

```haskell
let
  -- if the queue is closed, then return what we have already got;
  -- otherwise, retry until we can read an item from it.
  readItems xs =
    readTBMQueue q ≫= \case
      Nothing → return xs
      Just x → go (x:xs)

  -- retry until the timeout fires
  timeout = readTVar timedOutVar ≫= check

  go xs
    -- if we have got enough items, return immediately
    | length xs ≥ n = return xs
    -- either try reading from the queue, or, unless the queue is empty, 
    -- check we have run out of time
    | otherwise = readItems xs <|> if null xs then retry else xs <$ timeOut
```

In `go` first we try if we can read an item. If that fails (queue is neither closed or there is an item) we
try either

*   do nothing in case the we haven't yet read any item at all or
*   we check the timeout, if we have already read at least one item.

So, we block until more items get added to the queue only if

*   we haven't read any item at all,
*   we read some items and the timeout triggered (either before or after reading the items), or
*   we read more as much items as requested.

This means the only case an empty list is returned is when the queue is closed. Otherwise an non-empty list is returned precisely if either the timeout has triggered (possibly in the past) or we got as many items as requested.

I think that's the desired behavior.